### PR TITLE
chore: refactor resource dockerfile

### PR DESCRIPTION
**Pull request by dduportal**
While working on #390 and #412, the need to refactor the `dockerfile` resource was raised.

This PR does not change the functionnality of the resource: it's why the label says "chore", except for the resulting commit messages (that were not really correct).

It introduces the following changes:

- Fix the example to allow end to end testing from the root of the repository
- Apply the same conventions as other refactored resources:
  - Define a `Spec` type on the package with only the public attributes
  - Add a `New` initializer method
- Improve Developer experience:
  - Full unit test with mocked injected dependencies (no more file to be read/written on the FS while running tests!)
  - Increased test coverage that helped cleaning up unused struct attributes

## Test

To test this pull request, you can run the following commands:

```shell
go build -o ./dist/updatecli && ./dist/updatecli diff --config examples/updatecli.generic/dockerfile
```

### CODE REVIEW

File: pkg/core/pipeline/condition/main.go [hunk 1]
```diff
@@ -154,14 +154,16 @@ func Unmarshal(condition *Condition) (conditioner Conditioner, err error) {
 		}
 
 	case "dockerfile":
-		d := dockerfile.Dockerfile{}
+		var sourceSpec dockerfile.Spec
```
Comment (olblak) [inline pkg/core/pipeline/condition/main.go:157]: Shouldn't it be named "conditionSpec"?

File: pkg/core/pipeline/target/main.go [hunk 2]
```diff
@@ -85,15 +85,17 @@ func Unmarshal(target *Target) (targeter Targeter, err error) {
 		targeter = &ch
 
 	case "dockerfile":
-		d := dockerfile.Dockerfile{}
+		sourceSpec := dockerfile.Spec{}
```
Comment (olblak) [inline pkg/core/pipeline/target/main.go:88]: Shouldn't it be targetSpec?

File: pkg/core/pipeline/condition/main.go (hunk 1 above)
Comment (dduportal) [reply pkg/core/pipeline/condition/main.go]: Good catch! PR update (and rebased)

File: pkg/core/pipeline/target/main.go (hunk 2 above)
Comment (dduportal) [reply pkg/core/pipeline/target/main.go]: Good catch! PR update (and rebased)

**Review by olblak — APPROVED** [review]
It looks good to me
🚢

**Pull request merged**